django/db/models/base.py

```diff
@@ -538,6 +542,18 @@ def __hash__(self):
             raise TypeError("Model instances without primary key value are unhashable")
         return hash(self.pk)
 
+    def __copy__(self):
+        # We need to avoid hitting __reduce__, so define this
+        # slightly weird copy construct.
+        obj = Empty()
+        obj.__class__ = self.__class__
+        obj.__dict__ = self.__dict__.copy()
+
+        obj._state = copy.copy(obj._state)
+        obj._state.fields_cache = {}
+
+        return obj
```

Comment:
Do we need to use `Empty`? We will not hit `__reduce__()` with:
```python
def __copy__(self):
    obj = type(self)()
    obj.__dict__ = self.__dict__.copy()
    obj._state = copy.copy(obj._state)
    obj._state.fields_cache = {}
```

Reply:
FYI I following the approach taken in `Field.__copy__` (django/db/models/fields/__init__.py:542)

I am a bit hesitant to do that as it will have the following effects:
* pre_init/post_init signals will be emitted
* Field initialisation will occur

Both will change the behavior of model copying in a new way and will likely reduce performance without providing any benefit. Atleast that is my understanding of the code atm.

Seeing as I am not very familiar with the internal of the ORM I would defer this decision to anyone that is.

Reply:
Could you elaborate on _why_ you're trying to avoid calling `__reduce__`? That's what `copy.copy` was doing on Django 1.11 so that would be a change in behavior

https://github.com/django/django/blob/c669cf279ae7b3e02a61db4fb077030a4db80e4f/django/db/models/base.py#L618-L622

https://github.com/python/cpython/blob/e27a51c11e10d5df79b3e48dc3e7bfedfad5a794/Lib/copy.py#L86-L98

Reply:
I was blindly following `Field`'s approach which I now see is incorrect. Thanks for noticing that.

The new implementation adjusts the state provided to `__reduce__`. 

Thoughts?

Reply:
Looking into the build errors still.